codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,252 @@
+    private final WriteVisitor writeVisitor = new WriteVisitor();
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final int minAllocationChunk;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        this(connection, DEFAULT_MIN_ALLOCATION_CHUNK);
+    }
+
+    public UniformStreamByteDistributor(Http2Connection connection, int minAllocationChunk) {
+        this.connection = checkNotNull(connection, "connection");
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        State connectionState = new State(connectionStream);
+        connectionStream.setProperty(stateKey, connectionState);
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                State state = new State(stream);
+                stream.setProperty(stateKey, state);
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // Allocate the bytes to the streams.
+        long allocated = allocateBytes(maxBytes);
+
+        long unallocatedStreamableBytes = totalStreamableBytes - allocated;
+
+        // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
+        writeVisitor.writeAllocatedBytes(writer);
+
+        return unallocatedStreamableBytes > 0;
+    }
+
+    /**
+     * Allocates the given bytes across the streams.
+     *
+     * @param maxBytes The maximum number of bytes to allocate.
+     * @return the total number of bytes actually allocated.
+     */
+    private long allocateBytes(int maxBytes) {
+        if (queue.isEmpty() || maxBytes <= 0) {
+            return 0L;
+        }
+
+        long allocated = 0;
+        long unallocated = totalStreamableBytes;
+        int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+        while (maxBytes > 0 && !queue.isEmpty()) {
+            //int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+
+            // Remove the head of the queue.
+            State state = queue.remove();
```

Comment:
`remove()` -> `removeFirst()`?

We are being very particular about where we insert from using addLast and addFirst, perhaps we should do the same with removal?